src/Version10/Util/JUnitReportAppender.php

```diff
@@ -0,0 +1,125 @@
+                continue;
+            }
+            $this->walkTestSuite($report, $childNode);
+        }
+    }
+
+    private function walkTestSuite(ToolReportInterface $report, DOMElement $testsuite): void
+    {
+        foreach ($testsuite->childNodes as $childNode) {
+            if (!$childNode instanceof DOMElement) {
+                continue;
+            }
+
+            switch ($childNode->nodeName) {
+                case 'testsuite':
+                    // We are only interested in errors.
+                    if (!$this->wantToProcessTestSuite($childNode)) {
+                        break;
+                    }
+                    $this->walkTestSuite($report, $childNode);
+                    break;
+                case 'testcase':
+                    $this->walkTestCase($report, $childNode);
+            }
+        }
+    }
+
+    private function wantToProcessTestSuite(DOMElement $testsuite): bool
+    {
+        // We are only interested in errors, failures and the like.
+        if (0 !== $this->getIntXmlAttribute($testsuite, 'errors')) {
+            return true;
+        }
+        if (0 !== $this->getIntXmlAttribute($testsuite, 'failures')) {
+            return true;
+        }
+        if (0 !== $this->getIntXmlAttribute($testsuite, 'warnings')) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private function walkTestCase(ToolReportInterface $report, DOMElement $testCase): void
+    {
+        /*
+         * <testcase> has the following attributes:
+         *
+         * - name: name of test
+         * - class: FQDN of the test class
+         * - classname: dot separated FQDN of the class
+         * - file: file the test class is declared in (absolute path)
+         * - line: line number of failure
+         * - assertions: assertion count up until error
+         * - time: time test has consumed
+         */
+        foreach ($testCase->childNodes as $childNode) {
+            if (!$childNode instanceof DOMElement) {
+                continue;
+            }
+
+            switch ($childNode->nodeName) {
+                case 'error':
+                case 'failure':
+                    $severity = ToolReportInterface::SEVERITY_ERROR;
+                    break;
+                case 'warning':
+                    $severity = ToolReportInterface::SEVERITY_WARNING;
+                    break;
+                default:
+                    // FIXME: remove this.
+                    throw new \RuntimeException('Node name unknown: ' . $childNode->nodeName);
+            }
+            /*
+             * <failure> has the following attributes:
+             * - type: name of exception
+             *
+             * Text content holds error message.
+             */
+            $report->addDiagnostic(
+                $severity,
+                $this->stripRootDir($childNode->nodeValue),
+                $this->stripRootDir($testCase->getAttribute('file')),
```

Comment:
Phpspec omits the `file` attribute but provides the `classname` attribute. I'd like to resolve the file name but I'm not sure how as the autoload file is not loaded in the context, so the reflection API does not work. @discordier Do you have any idea?

Should we add a classname attribute to our `diagnostice` node as well?

Reply:
Having a class in the diagnostics might come in handy.
However, I think this should rather be an array of class names as there might be multiple classes affected (as we have with files and ranges).